design/token-exchange-spi.md

```diff
@@ -0,0 +1,86 @@
+# Token Exchange SPI
+
+## Abstract
+This document outlines the design of the proposed Token Exchange SPI for Keycloak (TODO: JIRA link).
+## Rationale
+Keycloak does implement the OAuth 2.0 Token Exchange ([RFC 8693](https://tools.ietf.org/html/rfc8693)), but does that in a peculiar way (Securing Applications and Services Guide, [7. Token Exchange](https://www.keycloak.org/docs/latest/securing_apps/index.html#_token-exchange)):
+> Token exchange in Keycloak is a very loose implementation of the OAuth Token Exchange specification at the IETF. We have extended it a little, ignored some of it, and loosely interpreted other parts of the specification.
+
+Current implementation has the following limitations and shortcomings:
+* The following (optional) parameters are not supported:
+	* `resource`
```

Comment:
What are your plans about the `resource` parameter? Isn't the audience enough and potentially easier to handle because it is a "logical" name of the targeted service?

I'm wondering if you have some specific requirement to choose the proper audience/target service based on a specific URL to where the token should be created for.

Reply:
The plan is to implement all of the parameters supported by the spec, but by delegating the values to the registered Providers so that they can inspect them and elect to handle this exchange. From this perspective, the `resource` parameter is just one of the options for a Provider to use to determine if it should be the one to handle this exchange request.

Reply:
I see. I would say that using the `resource` like that looks too fine-grained? I would just use both `audience` and `resource` parameters to influence how tokens should be issued, and not as a filter for choosing a provider?

Reply:
I guess the purpose of the matching algorithm is to provide a simple way for one of the Providers to decide to be the one that handles the exchange. They are free to use the provided parameters in any way they deem appropriate - if they want to register an interest in a URL or a logical name, and are able to work with such, then I'd say it's fair game and there's no reason to limit the request parameters that a Provider can make use of.